### PR TITLE
Ads Confirmations

### PR DESCRIPTION
**Pull request by tmancey**
fixes https://github.com/brave/brave-browser/issues/3117
fixes https://github.com/brave/brave-browser/issues/2830
fixes https://github.com/brave/brave-browser/issues/2912
fixes https://github.com/brave/brave-browser/issues/2910
fixes https://github.com/brave/brave-browser/issues/2909
fixes https://github.com/brave/brave-browser/issues/2908
fixes https://github.com/brave/brave-browser/issues/2907
fixes https://github.com/brave/brave-browser/issues/2906
requires https://github.com/brave/brave-browser/pull/3406
requires https://github.com/brave-intl/bat-native-confirmations/pull/67

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Windows
  - [x] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [x] Windows
  - [x] macOS
  - [x] Linux
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: components/brave_ads/browser/ads_service_impl.h [hunk 1]
```diff
@@ -17,6 +17,8 @@
 #include "bat/ads/ads_client.h"
 #include "brave/components/brave_ads/browser/ads_service.h"
 #include "brave/components/brave_ads/browser/background_helper.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
```
Comment (bridiver) [inline components/brave_ads/browser/ads_service_impl.h]: these headers should go in the .cc file - there is already a forward decl below for RewardsService

File: components/brave_rewards/browser/rewards_notification_service_impl.cc [hunk 2]
```diff
@@ -297,14 +297,29 @@ void RewardsNotificationServiceImpl::OnGetAllNotifications(
   TriggerOnGetAllNotifications(rewards_notifications_list);
 }
 
+std::string RewardsNotificationServiceImpl::GetGrantIdPrefix(
+    const std::string& grant_type) {
+  std::string prefix = grant_type == "ugp"
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_notification_service_impl.cc]: grant_type == "ugp" should be a method to remove duplication

File: components/brave_rewards/browser/rewards_service.h [hunk 3]
```diff
@@ -21,6 +21,11 @@
 class PrefRegistrySimple;
 class Profile;
 
+namespace ads {
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service.h]: these are not needed

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 4]
```diff
@@ -182,6 +182,25 @@ class RewardsServiceImpl : public RewardsService,
 
   void GetAddressesForPaymentId(const GetAddressesCallback& callback) override;
 
+  void SetCatalogIssuers(const std::string& json) override;
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.h]: these should all be listed under `// LedgerClient impl`

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 5]
```diff
@@ -226,6 +245,7 @@ class RewardsServiceImpl : public RewardsService,
   void OnPublishersListSaved(ledger::LedgerCallbackHandler* handler,
                              bool success);
   void OnTimer(uint32_t timer_id);
+  void TriggerOnContentSiteUpdated();
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.h]: there's no implementation for this

File: components/brave_rewards/browser/rewards_notification_service_impl.cc (hunk 2 above)
Comment (ryanml) [reply components/brave_rewards/browser/rewards_notification_service_impl.cc]: done

File: components/brave_rewards/browser/rewards_service_impl.h (hunk 5 above)
Comment (tmancey) [reply components/brave_rewards/browser/rewards_service_impl.h]: @ryanml One for you?

Comment (ryanml) [reply components/brave_rewards/browser/rewards_service_impl.h]: @tmancey must be a mistake as I did not work with such a method, will remove

Comment (ryanml) [reply components/brave_rewards/browser/rewards_service_impl.h]: done

**Comment by NejcZdovc** [comment]
@ryanml when everything is done from UI perspective please let me know so that I can do a review of it. Please do not merge until I do that. Thank you

File: components/brave_rewards/browser/rewards_service_impl.h (hunk 4 above)
Comment (tmancey) [reply components/brave_rewards/browser/rewards_service_impl.h]: Just checked, and it is now listed under so I gather you made this change

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 6]
```diff
@@ -800,9 +829,12 @@ void RewardsServiceImpl::OnGrantFinish(ledger::Result result,
     if (!Connected())
       return;
 
+    int report_type = grant.type == "ads"
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: This one should be `ledger::ReportType`

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 7]
```diff
@@ -1314,12 +1347,13 @@ void RewardsServiceImpl::TriggerOnRecoverWallet(ledger::Result result,
     observer.OnRecoverWallet(this, result, balance, newGrants);
 }
 
-void RewardsServiceImpl::SolveGrantCaptcha(const std::string& solution) const {
+void RewardsServiceImpl::SolveGrantCaptcha(const std::string& solution,
+                                         const std::string& promotionId) const {
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: nit: alignment is off

File: components/definitions/rewardsExtensions.d.ts [hunk 8]
```diff
@@ -51,6 +52,13 @@ declare namespace RewardsExtension {
     captcha?: string
     hint?: string
     status?: GrantStatus
+    type?: string
+  }
+
+  export interface GrantResponse {
```
Comment (NejcZdovc) [inline components/definitions/rewardsExtensions.d.ts]: why do we need `GrantResponse` we have `GrantInfo`

File: components/brave_rewards/resources/ui/reducers/grant_reducer.ts [hunk 9]
```diff
@@ -5,32 +5,44 @@
 import { Reducer } from 'redux'
 
 // Constant
+// Temporary (ryanml)
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/ui/reducers/grant_reducer.ts]: what does this comment means?

File: vendor/bat-native-ledger/src/bat/ledger/ledger.cc [hunk 10]
```diff
@@ -537,14 +538,16 @@ bool Grant::loadFromJson(const std::string& json) {
     error = !(d.HasMember("altcurrency") && d["altcurrency"].IsString() &&
         d.HasMember("probi") && d["probi"].IsString() &&
         d.HasMember("promotionId") && d["promotionId"].IsString() &&
-        d.HasMember("expiryTime") && d["expiryTime"].IsUint64());
+        d.HasMember("expiryTime") && d["expiryTime"].IsUint64()) &&
+        d.HasMember("type") && d["type"].IsString();
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/ledger.cc]: we need to move this one out as old states will not have type and we will not load it correctly. So it needs to be a separate if check that defaults to "".

Comment (ryanml) [reply vendor/bat-native-ledger/src/bat/ledger/ledger.cc]: @NejcZdovc updated

**Comment by NejcZdovc** [comment]
**Notification for grant is not cleared when you claim is complete**
- enable rewards via panel
- click claim
- finish grant
- click ok
- notification for token is still there


cc @ryanml

**Pull request merged**